core/src/main/java/org/apache/iceberg/io/OutputFileFactory.java

```diff
@@ -95,25 +100,71 @@ public OutputFileFactory(PartitionSpec spec, FileFormat format, LocationProvider
     this.operationId = operationId;
   }
 
+  public static Builder builderFor(Table table, FileFormat format, int partitionId, long taskId) {
+    return new Builder(table, format, partitionId, taskId);
+  }
+
   private String generateFilename() {
     return format.addExtension(
         String.format("%05d-%d-%s-%05d", partitionId, taskId, operationId, fileCount.incrementAndGet()));
   }
 
   /**
-   * Generates EncryptedOutputFile for UnpartitionedWriter.
+   * Generates an {@link EncryptedOutputFile} for unpartitioned writes.
    */
   public EncryptedOutputFile newOutputFile() {
     OutputFile file = io.newOutputFile(locations.newDataLocation(generateFilename()));
     return encryptionManager.encrypt(file);
   }
 
   /**
-   * Generates EncryptedOutputFile for PartitionedWriter.
+   * Generates an {@link EncryptedOutputFile} for partitioned writes in the default spec.
    */
   public EncryptedOutputFile newOutputFile(StructLike partition) {
+    return newOutputFile(defaultSpec, partition);
+  }
+
+  /**
+   * Generates an {@link EncryptedOutputFile} for partitioned writes in a given spec.
+   */
+  public EncryptedOutputFile newOutputFile(PartitionSpec spec, StructLike partition) {
```

Comment:
nit: wondering if there are tests covering this method and also the scenario writing to multiple specs in a single transaction?